include/alpaka/mem/buf/SetKernel.hpp

```diff
@@ -0,0 +1,80 @@
+/* Copyright 2020 Jeffrey Kelling
+ *
+ * This file is part of Alpaka.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#pragma once
+
+#include <alpaka/mem/buf/Traits.hpp>
+#include <alpaka/idx/Traits.hpp>
+#include <alpaka/idx/Accessors.hpp>
+#include <alpaka/idx/MapIdx.hpp>
+
+namespace alpaka
+{
+    namespace mem
+    {
+        namespace view
+        {
+            //#############################################################################
+            //! any device ND memory set kernel.
+            class MemSetKernel
+            {
+            public:
+                //-----------------------------------------------------------------------------
+                //! The kernel entry point.
+                //!
+                //! All but the last element of threadElemExtent must be one.
+                //!
+                //! \tparam TAcc The accelerator environment to be executed on.
+                //! \tparam TElem element type.
+                //! \tparam TExtent extent type.
+                //! \param acc The accelerator to be executed on.
+                //! \param val value to set.
+                //! \param dst target mem ptr.
+                //! \param extent area to set.
+                ALPAKA_NO_HOST_ACC_WARNING
+                template<
+                    typename TAcc,
+                    typename TElem,
+                    typename TExtent,
+                    typename TPitch>
+                ALPAKA_FN_ACC auto operator()(
+                    TAcc const & acc,
+                    TElem const val,
+                    TElem * dst,
+                    TExtent extent,
+                    TPitch pitch) const
+                -> void
+                {
+                    using Idx = typename idx::traits::IdxType<TExtent>::type;
+                    auto const gridThreadIdx(alpaka::idx::getIdx<alpaka::Grid, alpaka::Threads>(acc));
+                    auto const threadElemExtent(alpaka::workdiv::getWorkDiv<alpaka::Thread, alpaka::Elems>(acc));
+                    auto const idxThreadFirstElem = idx::getIdxThreadFirstElem(acc, gridThreadIdx, threadElemExtent);
+                    auto idx = idx::mapIdxPitch<1u, dim::Dim<TAcc>::value>(idxThreadFirstElem, pitch)[0];
+                    constexpr auto lastDim = dim::Dim<TAcc>::value - 1;
+                    const auto lastIdx = idx +
+                        std::min(threadElemExtent[lastDim], static_cast<Idx>(extent[lastDim]-idxThreadFirstElem[lastDim]));
+
+                    if ([&idxThreadFirstElem, &extent](){
+                            for(auto i = 0u; i < dim::Dim<TAcc>::value; ++i)
+                                if(idxThreadFirstElem[i] >= extent[i])
+                                    return false;
+                            return true;
+                        }())
```

Comment:
May we be able to express this via an element-wise vector comparison and a logical fold-expression?
`if((idxThreadFirstElem < extent).foldrAll(std::logical_and<bool>()))`?

Reply:
I think we can.

According to cppreference.com, `std::logical_and` "Effectively calls operator&& on type T", thus with `foldr` calling `f(t0, foldr(f, t1, ts...))` this should short-circuit just like the loop, right? (Not important on GPU, but somewhat useful on CPU).

Reply:
I am not sure. This is also what I thought about and my conclusion was that there will be no short circuiting because it is a function call and not an inline &&. I would still prefer the shorter code here.